tensorflow_io/avro/20190307-avro-dataset.md

```diff
@@ -0,0 +1,299 @@
+  - Nested records
+  - Namespaces
+  - Branches in unions -- will resolve them but will use strict typing which will render them useless whenever an attribute has multiple types or whenever a null is encountered. Note, TensorFlow uses strict typing for it&#39;s tensors.
+- An API for filtering in arrays
+- The mapping of Avro enumerations to TensorFlow strings carrying the symbol name
+- The mapping of Avro fixed types to TensorFlow strings
+- The same operating systems as supported by TensorFlow -- Note, that this only accounts for the development of the avro dataset and not it&#39;s dependencies
+- Python 2.x and 3.x following TensorFlow&#39;s requirements
+- Thread safety; Note that TensorFlow&#39;s  dataset API may use multiple threads
+
+| **Avro primitive type** | **TensorFlow primitive type** |
+| --- | --- |
+| null: no value | If provided default is assumed otherwise fail |
+| boolean: a binary value | tf.bool |
+| int: 32-bit signed integer | tf.int32 |
+| long: 64-bit signed integer | tf.in64 |
+| float: single precision 32-bit IEEE floating point number | tf.float32 |
+| double: double precision 64-bit IEEE floating point number | tf.float64 |
+| fixed: a fixed length byte sequence | tf.string |
+| bytes: a sequence of 8-bit unsigned bytes | tf.string |
+| string: unicode character sequence | tf.string |
+| enum: enumeration | tf.string |
+
+**Table 1** lists the mapping of Avro primitive types to TensorFlow primitive types.
+
+The Avro dataset shall NOT support
+
+- --Type promotion or type coercion when mapping from Avro types to TensorFlow types, instead we use strict typing here
+- --Recursive type definitions in avro
+- --The compile and distribution of TensorFlow external tools that are necessary; e.g. libavro.so. It assumed that the user will compile and distribute this library separately.
+
+## Design Details
+
+The design shall leverage existing software as follows
+
+- For the API leverage the tf.dataset natively and corresponding python classes
+- Leverage the TensorFlow filesystem natively. Note, that currently meta information for file systems are mostly provided through environment variables (e.g. HDFS). For more information follow the special interest group (SIG) tensorflow.io and these links: [Hadoop](https://www.tensorflow.org/deploy/hadoop) and [gfile](https://www.tensorflow.org/api_docs/python/tf/gfile) .
+- Leverage Avro&#39;s c API. Note, the c++ API for Avro depends on boost which the TensorFlow author&#39;s do not want to depend on nor maintain
+- Leverage TensorFlow&#39;s build system for dependency management and build
+- Leverages Google&#39;s regex re2 package to parse user-provided strings to resolve record nestings, namespaces, array and map indexing
+- Follow google&#39;s c++ style guide and google&#39;s python style guide
+- We develop against Avro 1.8.2 with bugfixes for jansson build -- which is the latest stable release; see [avroc](https://github.com/apache/avro/tree/release-1.8.2/lang/c)
```

Comment:
I'm not familiar with the Avro community, but it looks like the last Avro-C release was quite a while ago. Is the project still active and maintained?  Is Avro C++ built on top of C or is it a complete separate API?

Reply:
It's a complete separate API and implementation; see
c API https://github.com/apache/avro/tree/master/lang/c/src/avro
c++ API https://github.com/apache/avro/tree/master/lang/c%2B%2B/api

Overall
c++ API 
+ user friendly 
+ more active commits
- requires boost as dependency
c API
+ no additional dependencies
+ easy build (although snappy may give some extra trouble)
- not so user-friendly for high-level programmers (scala, java, c++)

Reply:
+1 on C++ as long as the dependencies don't pose a problem. I think snappy might be ok as it is also used by Arrow/ParquetCpp.